contracts/src/interfaces/IGateway.sol

```diff
@@ -110,4 +171,56 @@ interface IGateway {
         uint128 destinationFee,
         uint128 amount
     ) external payable;
+
+    // V2
+
+    // Submit a message for verification and dispatch
+    function v2_submit(
+        InboundMessageV2 calldata message,
+        bytes32[] calldata leafProof,
+        Verification.Proof calldata headerProof,
+        bytes32 rewardAddress
+    ) external;
+
+    // Send an XCM with arbitrary assets to Polkadot Asset Hub
+    //
+    // Params:
+    //   * `xcm` (bytes): SCALE-encoded VersionedXcm message
+    //   * `assets` (bytes[]): Array of asset specs, constrained to maximum of eight.
+    //
+    // Supported asset specs:
+    // * ERC20: abi.encode(0, tokenAddress, value)
+    //
+    // On Asset Hub, the assets will be received into the assets holding register.
+    //
+    // The `xcm` should contain the necessary instructions to:
+    // 1. Pay XCM execution fees for `xcm`, either from assets in holding,
+    //    or from the sovereign account of `msg.sender`.
+    // 2. Handle the assets in holding, either depositing them into
+    //    some account, or forwarding them to another destination.
+    //
+    // To incentivize message delivery, some amount of ether must be passed and should
+    // at least cover the total cost of delivery to Polkadot. This ether be sent across
+    // the bridge as WETH, and given to the relayer as compensation and incentivization.
+    //
+    function v2_sendMessage(
+        bytes calldata xcm,
+        bytes[] calldata assets,
+        bytes calldata claimer
+    ) external payable;
+
+    // Register Ethereum-native token on AHP, using `xcmFeeAHP` of `msg.value`
+    // to pay for execution on AHP.
+    function v2_registerToken(address token, uint128 xcmFeeAHP) external payable;
+
+    // Register Ethereum-native token on AHK, using `xcmFeeAHP` and `xcmFeeAHK`
+    // of `msg.value` to pay for execution on AHP and AHK respectively.
+    function v2_registerTokenOnKusama(
+        address token,
+        uint128 xcmFeeAHP,
+        uint128 xcmFeeAHK
+    ) external payable;
+
+    // Check if an inbound message was previously accepted and dispatched
+    function v2_isDispatched(uint64 nonce) external returns (bool);
```

Comment:
Should add `view` keyword here.

https://github.com/Snowfork/snowbridge/pull/1321/commits/2e8d519e9c35b6f7cc41eb4a4d159af491a8c0e0